src/Sarif.UnitTests/Core/FileDataTests.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Security.AccessControl;
+using System.Text;
+using FluentAssertions;
+
+using Microsoft.CodeAnalysis.Sarif.Writers;
+
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class FileDataTests
+    {
+        [Fact]
+        public void FileData_Create_NullUri()
+        {
+            Action fileDataCreation = () => { FileData.Create(null, LoggingOptions.None); };
+
+            fileDataCreation.ShouldThrow<ArgumentNullException>();
+        }
+
+        [Fact]
+        public void FileData_ComputeHashes()
+        {
+            string filePath = Path.GetTempFileName();
+            string fileContents = Guid.NewGuid().ToString();
+            Uri uri = new Uri(filePath);
+
+            try
+            {
+                File.WriteAllText(filePath, fileContents);
+                FileData fileData = FileData.Create(uri, LoggingOptions.ComputeFileHashes);
+                fileData.Uri.Should().Be(uri.ToString());
+                HashData hashes = HashUtilities.ComputeHashes(filePath);
+                fileData.Hashes.Count.Should().Be(3);
+
+                foreach (Hash hash in fileData.Hashes)
+                {
+                    switch (hash.Algorithm)
+                    {
+                        case AlgorithmKind.MD5: { hash.Value.Should().Be(hashes.MD5); break; }
+                        case AlgorithmKind.Sha1: { hash.Value.Should().Be(hashes.Sha1); break; }
+                        case AlgorithmKind.Sha256: { hash.Value.Should().Be(hashes.Sha256); break; }
+                    }
+                }
+            }
+            finally
+            {
+                if (File.Exists(filePath)) { File.Delete(filePath); }
+            }
+        }
+
+        [Fact]
+        public void FileData_PersistFileContents()
```

Comment:
Need a test for binary files. The spec says those are just plain base64 encoded. The idea was that the tool creating the log file presumably knows whether it's analyzing a source file or a binary file.

I think what has to happen is that `FileData.Create` takes an additional argument specifying whether the file is text or binary. But how will we know? We can assume binary unless we have a MIME type?

Reply:
Oh! I just looked at the product code (I always look at the tests first :-)) -- and you have the MIME type available.

---
In reply to: [136158851](https://github.com/Microsoft/sarif-sdk/pull/612#discussion_r136158851) [](ancestors = 136158851)

Reply:
you are now proposing a useful update to the Create implementation, deferring. I will do this next, add mocks, and also convert everything to xunit

---
In reply to: [136158851](https://github.com/Microsoft/sarif-sdk/pull/612#discussion_r136158851) [](ancestors = 136158851)

Reply:
actually probably should fix now, since we're releasing.

---
In reply to: [136160339](https://github.com/Microsoft/sarif-sdk/pull/612#discussion_r136160339) [](ancestors = 136160339,136158851)

Reply:
This isn't a "useful update". The current behavior violates the spec.

---
In reply to: [136160339](https://github.com/Microsoft/sarif-sdk/pull/612#discussion_r136160339) [](ancestors = 136160339,136158851)

Reply:
right, precisely why i reversed my position. see this update.

---
In reply to: [136160846](https://github.com/Microsoft/sarif-sdk/pull/612#discussion_r136160846) [](ancestors = 136160846,136160339,136158851)